Make HasValue virtual in wxDataViewModel and have implementations use it

jmoraleda opened the pull request:
This addresses issue https://trac.wxwidgets.org/ticket/18724

jmoraleda committed:
Make HasValue virtual in wxDataViewModel and have implementations use it. This addresses issue https://trac.wxwidgets.org/ticket/18724

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2394,15 +2394,17 @@ void wxDataViewMainWindow::OnPaint( wxPaintEvent &WXUNUSED(event) )
-                    // If this is container node without columns, render full-row focus:
-                    if ( !IsList() )
+                    // If this is an item with a single column, render full-row focus:
+                    numColsWithValue = 0;
```
This is what seems to result in build errors on Travis CI, this variable is not declared.

jmoraleda replied:
Thank you. Fixed.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2394,15 +2394,17 @@ void wxDataViewMainWindow::OnPaint( wxPaintEvent &WXUNUSED(event) )
-                            renderColumnFocus = false;
+                        if ( model->HasValue(item, i) )
+                            numColsWithValue++;
+                        if ( numColsWithValue > 1 )
+                            break;
                     }
+                    renderColumnFocus = numColsWithValue<2;
```
I'd just replace this with
```suggestion
                        if ( numColsWithValue > 1 )
                        {
                            renderColumnFocus = false;
                            break;
                        }
                    }
```
IMHO it's a bit clearer.

jmoraleda replied:
By creating a function `HasItemMultipleValues` I have moved the new code out of here and left the logic here virtually identical to the original one. See commit 20243534ca9853a2d13a678379fe0a46398f01cf

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2394,15 +2394,17 @@ void wxDataViewMainWindow::OnPaint( wxPaintEvent &WXUNUSED(event) )
+                    // If this is an item with a single column, render full-row focus:
+                    numColsWithValue = 0;
+                    unsigned int cols = GetOwner()->GetColumnCount();
+                    for ( unsigned int i = 0; i < cols; i++ )
```
I am not sure why did we have to check for `!IsList()` before, but are you sure this is not needed any longer?

jmoraleda replied:
By creating a function HasItemMultipleValues I have moved the new code out of here and left the logic here virtually identical to the original one. See commit 20243534ca9853a2d13a678379fe0a46398f01cf

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -4500,16 +4496,27 @@ bool wxDataViewMainWindow::TryAdvanceCurrentColumn(wxDataViewTreeNode *node, wxK
-        // navigation shouldn't work in branch nodes without other columns:
-        if ( node->HasChildren() && !GetModel()->HasContainerColumns(node->GetItem()) )
+        // navigation shouldn't work in nodes with fewer than two columns
+        numColsWithValue = 0;
+        unsigned int cols = GetOwner()->GetColumnCount();
+        for ( unsigned int i = 0; i < cols; i++ )
+            if ( GetModel()->HasValue(node->GetItem(), i) )
+                numColsWithValue++;
+        if (numColsWithValue<2)
```
As this fragment is being used at least twice, it would be preferable to add a function, e.g. `DoesColumnHaveSingleItemOnly()` (but please feel free to use a better name if you see one) and call it from both places.

jmoraleda replied:
Done. By creating a function HasItemMultipleValues I have moved the new code out of here and left the logic here virtually identical to the original one. See commit 20243534ca9853a2d13a678379fe0a46398f01cf.

vadz reviewed (commented):
Thanks, globally this looks good, but the CI build failures need to be corrected, of course.

Also, I wonder if we can really override `HasValue()` to return false for the column 0. It seems like it would be pretty weird to not have anything in the "expander" column. But why not...

As for the keyboard navigation logic, it's indeed not totally obvious. The best would really be to provide some way of testing it easily in the dataview sample, i.e. override `HasValue()` for demonstration purposes in one of the models defined there. Could you please add this?

TIA!

jmoraleda committed:
Bug fix. Ensure travis builds without errors on all platforms.

jmoraleda committed:
Consolidate common logic into a single function and revert original test for IsList() to minimize code changes.

jmoraleda committed:
Created method to find the first column that has a value instead of assuming that the expander column is the first one. Renamed method HasItemMultipleValues with IsItemMultivalued and made it const.

jmoraleda committed:
Added an example of overriding method HasValue in wxDataView to show how some items may not have values for some columns.

jmoraleda commented:
Hello Vadim, and thank you for all your feedback. I addressed all your comments. (See details for comment above). I also added a (trivial) example of overriding the `HasValue` method in in the dataview samples: I added a new 7th panel that is identical to the third panel, except that it overrides HasValue so that the diagonal elements don't have values. This in particular shows not having a value in the 0th column. On my linux debian with gtk3 the example behaves as I would expect, not displaying anything for the cells that don't have values.

Let me know what you think.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -910,6 +910,33 @@ class wxDataViewMainWindow: public wxWindow
+    {
+        unsigned int cols = GetOwner()->GetColumnCount();
+        const wxDataViewModel* model = GetModel();
+        for ( unsigned int i ; i < cols; i++ )
```
```suggestion
        for ( unsigned int i = 0; i < cols; i++ )
```
Unless I'm missing something, there is a bug here due to `i` not being initialized at all.

vadz replied:
Hmm, this got resolved, but the variable remained uninitialized. I'm adding ` = 0` here.

vadz reviewed (commented):
Thanks a lot for the updates, this looks very nice now (mostly speaking about the generic version code)!

However I think there is still a bug remaining, could you please fix this? TIA!

jmoraleda commented:
Thank you again for your reviews. They have been very helpful, especially since the generic code is not run in Linux, which is the only platform I have access to.

RobinD42 commented:
Just FYI, I've been testing this on OSX.  It looks like all is working as expected there.

jmoraleda commented:
Thank you @RobinD42 ! (FYI actually this started as a Phoenix issue https://github.com/wxWidgets/Phoenix/issues/1594 until I realized the issue was in wxWidgets. I have already tested that once `HasValue` becomes virtual it gets picked up by Phoenix without having to do anything else and can be subclassed in python and works as expected)

RobinD42 commented:
Yep, I've been testing the Python side of things as well.

vadz commented:
> Thank you again for your reviews. They have been very helpful, especially since the generic code is not run in Linux, which is the only platform I have access to.

FWIW I use a VM for testing wxMSW when working under Linux, but it's also possible to use Wine.

RobinD42 commented:
It looks like when sorting by clicking on a column header it is still calling `GetValue` for cells that `HasValue` will have returned returned False.

RobinD42 commented:
This seems to take care of it for me on OSX, I'm not sure if anything else is needed however.

```
diff --git a/src/common/datavcmn.cpp b/src/common/datavcmn.cpp
index 76073485e1..f9cb77f742 100644
--- a/src/common/datavcmn.cpp
+++ b/src/common/datavcmn.cpp
@@ -334,8 +334,8 @@ int wxDataViewModel::Compare( const wxDataViewItem &item1, const wxDataViewItem
                               unsigned int column, bool ascending ) const
 {
     wxVariant value1,value2;
-    GetValue( value1, item1, column );
-    GetValue( value2, item2, column );
+    HasValue(item1, column) ? GetValue( value1, item1, column ) : value1.MakeNull();
+    HasValue(item2, column) ? GetValue( value2, item1, column ) : value2.MakeNull();

     if (!ascending)
     {
```

jmoraleda committed:
Use null as value when sorting items in a DataView control by a column for which the item does not have a value.

jmoraleda commented:
Thank you Robin. Good catch! I just committed your proposed changes.

RobinD42 commented:
@vadz, does this need anything else before merging?

vadz commented:
No, I'll try to merge this soon, but I'd like to test it under MSW first and so I need to find some time to do it.

RobinD42 commented:
FWIW, my Python sample using `HasValue` works fine on MSW, GTK and OSX.

vadz commented on src/common/datavcmn.cpp:
```diff
@@ -334,8 +334,8 @@ int wxDataViewModel::Compare( const wxDataViewItem &item1, const wxDataViewItem
-    GetValue( value1, item1, column );
-    GetValue( value2, item2, column );
+    HasValue(item1, column) ? GetValue( value1, item1, column ) : value1.MakeNull();
+    HasValue(item2, column) ? GetValue( value2, item2, column ) : value2.MakeNull();
```
I don't see why do we need to call `MakeNull()` here, this seems to be redundant and using ternary operator without using its return value is rather strange, so I'm changing it to just use a simple `if`.

jmoraleda replied:
Thank you. Good catch.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -910,6 +910,33 @@ class wxDataViewMainWindow: public wxWindow
     bool DoItemChanged(const wxDataViewItem& item, int view_column);
 
+    // Return whether the item has more than one column that have values
+    bool IsItemMultivalued(const wxDataViewItem& item) const
```
Looking at the code, we always use `!IsItemMultivalued()`, so it seems better to rename this to `IsItemSingleValued()` and remove the negation.

jmoraleda replied:
I considered this alternative. I opted for the multivalued naming as being marginally more precise because it is theoretically possible for an item to have no values and an item with 0 values should be treated as one with a single value, rather than as one with more than one value. Let me know if you still want me to change this.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -4498,18 +4514,16 @@ bool wxDataViewMainWindow::TryAdvanceCurrentColumn(wxDataViewTreeNode *node, wxK
         {
-            m_currentCol = GetOwner()->GetColumnAt(0);
+            // find first column with value
+            m_currentCol = FindFirstColumnWithValue(node->GetItem());
```
This is a fatal bug as NULL `node` can be dereferenced here -- and this is exactly what happens if you just press a right arrow on the new page of the sample.

I've fixed this by restoring the old code for this case.

vadz committed:
Mention that wxDataViewModel::HasValue() became virtual in 3.1.4

This method existed since earlier versions, but couldn't be overridden
until now.

vadz committed:
Don't call wxVariant::MakeNull() unnecessary

Simplify recently added code in wxDataViewModel::Compare().

No real changes.

vadz committed:
Replace IsItemMultivalued() with IsItemSingleValued()

This allows to simplify logic when calling it by avoiding negation.

No real changes.

vadz committed:
Fix crash due to dereferencing null pointer in list mode

TryAdvanceCurrentColumn() is called with NULL node in this case, so
avoid dereferencing it.

vadz committed:
Minor formatting and code style fixes

Put braces around loop body.

Use const for local variables.

No real changes.

vadz reviewed (changes requested):
Sorry, but this needs more work in the generic version.

I'm pushing out my fixes that at least fix the crash, but I ran out of time for more and the keyboard behaviour is still wrong: navigation doesn't skip the empty cells, as it would be supposed to, and pressing Space when the cursor is on them toggles the checkboxes. I am not sure how is this supposed to work in the special "list" case, but it should be improved and, clearly, it must be tested. If you really can't use a MSW VM, please use Wine or, at the very least, hack your wx build to use the generic version of the control in wxGTK, but doing such non-trivial changes without testing is just not a good idea.

RobinD42 commented:
@jmoraleda, FYI, you can force the use of the generic DVC classes by using the `--disable-nativedvc` flag when running `configure`.

jmoraleda commented:
@RobinD42 Thank you for the tip. Very helpful. I will make time later today or tomorrow to use your suggestion to attempt to reproduce and fix the issues that @vadz is reporting with keyboard navigation.

jmoraleda committed:
A DVC cell that has no value is not editable

jmoraleda commented:
I was able to compile and run the generic code using ` --disable-nativedvc`. This beats relying on the travis-ci output for debugging and testing :smile: 

I was able to reproduce issues with keyboard and mouse navigation and selection in the generic component. I just pushed a commit that appears to fix for me the last remaining issue by ensuring that a cell with no value cannot be edited.

Thank you again @vadz and @RobinD42 for all your help.

vadz commented:
Sorry but the issue I pointed out is still there. Reproducing it is trivial: just go the last tab (the one added by this PR) in the sample, press `Down`, `Right`, `Right` and the cursor (selected) cell is still on the blank cell ("A2") and pressing `Space` toggles the checkbox in the first column.

jmoraleda commented:
Thank you again for your time on this.

The behavior of the dataview control in my system, including the sample, is different from what I think it must be in yours, which is why I have not noticed before the problem you report.

For me, individual cells are never visually selected in a dataview control, only whole rows are, regardless of whether I select using the keyboard, the mouse or a combination. This is true both for the native control and the generic control, in all tabs of the demo and my own applications, and as far as I can recall has always been so. Furthermore, as a sanity check, I just verified again that this behavior happens in the master branch as well, both with the native and generic control and is thus independent of this PR.

Can you shed light on whether this is a known side effect of being on linux/gtk, even when using the generic control; or am I missing some compilation option, or am I experiencing a separate bug?

As I mentioned, the visual issue is not related to this PR at all, but it is affecting my debugging abilities with respect to the issue you describe: I have tried using the keyboard to navigate cells assuming that individual cells are selected, even if there are no visual cues. I feel there is indeed some unexpected behavior of tab/arrow right/space with respect to what checkbox/radio box gets toggled, but it is hard to debug since I cannot see which cell is supposed to be selected after a particular set of arrow key presses.  
 
Just for reference, this is what the new tab looks with the native control with nothing selected
![image](https://user-images.githubusercontent.com/23492126/79697431-addd6000-8237-11ea-8dba-bb0aa69b8e01.png)
This is what the new tab looks like with the generic control with nothing selected
![image](https://user-images.githubusercontent.com/23492126/79701622-741a5280-8253-11ea-8cb3-9b8c6352afe6.png)

This is the generic with a row selected (since I cannot select an individual cell):
![image](https://user-images.githubusercontent.com/23492126/79701963-f99f0200-8255-11ea-988e-651ed9aeb108.png)

Mouse clicks work correctly in both native and generic control.

vadz commented:
Sorry, I didn't have time to test the generic version under Linux yet, but if it doesn't show the selected cell, it's indeed a bug. You don't have to fix it as part of this PR, of course, but if you really can't run MSW in VM nor Wine (although it's difficult to imagine what might be preventing you from doing the latter...), maybe fixing it could be a nice bonus side effect.

I do see the selected cell in the GTK version but it's _very_ difficult to see, there is just an almost invisible focus rectangle around the label for it. It probably depends on the theme being used, but with the default GTK 3 theme it's indeed quite useless -- the generic version focus appearance is not very accented, but GTK 3 is even worse. FWIW it's slightly better in GTK 2, where the focus rectangle is at least black and not white.

jmoraleda commented:
> I do see the selected cell in the GTK version but it's very difficult to see, there is just an almost invisible focus rectangle around the label for it.

 Wow! Now that you point this out, I can see it too! Indeed useless for real world usage and does not show in screen captures, but I will make it suffice for debugging. Thank you again.

jmoraleda committed:
Added dedicated command binding of HasPage tab of DVC sample so that the radio button column has always exactly one cell checked. Before this fix this tab was incorrectly bound to the same function that controlled the ListCtrl in the third tab.

jmoraleda committed:
In DVC, if current column was set by keyboard to something not editable and the user pressed Space/F2 then do not edit anything because focus is visually on that column and editing something else would be surprising. But if the current column was set by mouse to something not editable, treat the situation as if there was whole-row focus, because the mouse click could very well be targeted on the row rather than on an individual cell.

jmoraleda committed:
Consistent indentation in generic DVC

jmoraleda commented:
I was able to reproduce the issue of 
> press Down, Right, Right and the cursor (selected) cell is still on the blank cell ("A2") and pressing Space toggles the checkbox in the first column.

I believe it is fixed in the last version.

jmoraleda commented:
Note that the the last travis error is due to a timeout and I do not have permissions to trigger a retry. I believe the current version addresses all issues that have been brought up and may be ready for a pull.

jmoraleda commented:
I have read somewhere that closing and reopening will trigger a retry. Given it a try.

jmoraleda reviewed (commented):
I believe the current version addresses all issues that have been brought up and may be ready for a pull.

Note that the last commit triggered a transitory travis error due to a timeout. I am closing and reopening the PR to trigger a retry.

vadz commented:
Thanks, I feel like it would still be better if the keyboard focus skipped the value-less cells even when using a list model, but this is not critical and I'll merge it as is and leave any improvements to a subsequent PR.

Thanks again!
